### ISSUE TITLE
Multi Threading with Web Workers or Web Assembly?

### ISSUE THREAD

**Issue by Juliiiian** [Dezinater/RustWorldJS#2]
I was looking around, on how to make the decompression process faster. And my friend saw that you can use multi threading. It will be probably a lot of work to use web workers or even switch to use web assembly with threads.
It's just a question and I want to know your opinion.

**Comment by Dezinater** [comment]
I don't think it should be too bad to add multithreading. If you take a look at LZ4Writer it's actually using multithreading already to write the chunks. I only included it for the writing since when I was testing I noticed writing took considerably longer. 

But basically what the reader and writer are doing is just dealing with the header and sending the LZ4 data to the LZ4 library. So the processes for multithreading the reader should be very similar to the writer. If you'd like to try to implement this you can but I can also do it too if you'd like but it might just take some time cause I'm limited on free time now :(

Also if you do try to implement it make sure to use the 'web-workers' package I'm using in LZ4Writer! It's a polyfill to make workers work on both Node and Web.

I think if you wanted to go down a Web Assembly route it would basically be rewriting the LZ4 package that I use, I think that's where you'll gain the most performance. But I think threads should give a good enough boost.

**Comment by Juliiiian** [comment]
Nice to hear that you are already using it for the LZ4writer. Then I guess it shouldn't be that hard to use workers it for the LZ4Reader. I looked before into the LZ4Reader to get the size of the decompressed file, but never looked into the LZ4Writer. 
No need for you to waste your limited free time. I should be able to get it working.
Thank you for the answer :)

**Comment by Dezinater** [comment]
Awesome sounds good :)
I'll keep this issue open for now incase there's anymore questions or even new ideas/discussion related to this

**Comment by Juliiiian** [comment]
Yes I have a question, I worked through the code but I still don't know what the do while loop does in [LZ4Reader](https://github.com/Dezinater/RustWorldJS/blob/cd298f45e5af7421c90e370ab4080a6c7472538a/src/LZ4Reader.ts#LL86C43-L86C43)
And I quite dont understand what _buffer does or for what it is needed.

**Comment by Dezinater** [comment]
Since the LZ4 parts of this code are a direct conversion from the C# library ([as seen here](https://github.com/MiloszKrajewski/lz4net/blob/master/src/LZ4/LZ4Stream.cs#L274)) it might be kind of hard to tell what the original intention with some parts. Even in his code it says "skip empty block (shouldn't happen but...)" so its even more confusing why it's there. I'm not entirely sure why some parts are the way they are since I'm not too familiar with LZ4 and all of the special cases it might run in to. I tried to keep my code as similar to the original as I code so that it behaves in the same way. You can write it in a way that makes more sense but it is possible it could lead to some bugs we won't notice until it hits a special case in some map. 

Maybe they had this for streaming LZ4 instead of only just reading from file, then it might be possible that a block comes in as empty? I can't really think of any other reason why there would be a header but no data.

```js
AquireNextChunk() {
...
let originalLength = this.ReadVarInt();
...
this._bufferLength = originalLength;
...
} while (this._bufferLength == 0); // if the section was empty then repeat
```

For multithreading this I think you would basically just have to replace the decompress line with sending the data to the Worker to decompress. And then just handle how it comes back and gets put back together. For the Writer I just tag the message I send to the Worker with an index so I know where it goes in the output when I get the value back from the Worker

**Comment by Juliiiian** [comment]
I have a lot to do right now, and I'm not so motivated for the project where I need the lib. It should be finished like 70% a few bugs and some buffer errors. I will make a pull request in about 2 weeks when it should be finished.

**Comment by Dezinater** [comment]
That's alright take your time with it. It can be good to take a break and come back to it. What project are you making with this lib?

**Comment by Juliiiian** [comment]
My goal is that you can upload a rust.map file and then view it in the browser with three.js. Maybe even if this works, that you can login and just paste a seed and my server gets the .map file. And I have a question to the height map data, since you already made your own editor in c#. When I use the map data to construct a height map in three.js It's really spiky 
![picture](https://github.com/Dezinater/RustWorldJS/assets/89937721/34c9ea95-7af6-4c44-8f38-e3d97e027de8)
I guess it's because it's not normalized, but I usually don't do anything graphics related, so I have no clue on how to fix it.

**Comment by Dezinater** [comment]
Sounds like a cool project. You can link it here if you'd like after you're done with it. 

I'm not entirely sure why it's looking spiky like that but you are right it could be the normalization. You could try to normalize it and play around with random values for it. I think the map also may need to be offset possibly. It's been a while since I've worked on rust map stuff so I forget a bit. It does seem like I offset it in the C# code

 https://github.com/Dezinater/Rust-Map-Editor/blob/master/Assets/Scripts/MapIO.cs#L125

**Comment by Juliiiian** [comment]
Thank you, I will try that. I can link it tomorrow when I cleaned up this whole mess.

**Comment by Juliiiian** [comment]
I've invited you. It's written in svelte, so you might not understand a bit, but under src/lib/components/Render.svelte you can see how three JS stuff is done (highly inefficient). I need to improve a lot and find another way to create the mesh, but should be okay with web workers. I also included some comments, It will probably not work because right now I changed your lib to give me the finalSize of the file, but I made a PR for that quick fix.

**Comment by Dezinater** [comment]
I've looked at the code a bit but I didn't get a chance to run it yet. Did you figure out what was causing the height issues? If you need to normalize the values or something then maybe the lib needs to be changed so it's easier to read the height values instead.

**Comment by Juliiiian** [comment]
I will try it when I have time, btw I got a working version with web workers, but I have 2 issues that I need to fix. First is when compiling the .cts file extension is complaining because I have already initialized the variables like BLOCK_SIZE, but I will just rename them different. Is there a reason to use .cts? I don't know what it is doing. And the 2 issue is that the web worker example takes 2x as long on my machine than using the version without. But I think this should be an easy fix as well, I will merge when I have time to finish.

**Comment by Dezinater** [comment]
Weird that it's giving an error for having that variable initialized already. I had to use cts because of some Typescript/Webpack issues I believe. Honestly I'm thinking about just getting rid of the typescript. I'm making a library for another game and I'm for sure not going to use it for that. It causes too many problems for what it's worth. Even the creator of svelte wrote an article on why Typescript is bad. Typescript is amazing, but for libraries it would be good enough to just use JSDocs. What do you think? There would still be type checking with JSDocs but it wouldn't be strict and we'd be able to document how methods work/are intended to work.

I wonder why it would be slowing down so much. My only guess would be it's not actually running in parallel so it's basically the same as before but now it's just taking longer because it has to upload data to the worker? If you use the Chrome devtools and do a CPU, after you stop it you can go to the sources tab and it'll show you which lines took how long to execute

**Comment by Dezinater** [comment]
Weird that it's giving an error for having that variable initialized already. I had to use cts because of some Typescript/Webpack issues I believe. Honestly I'm thinking about just getting rid of the typescript. I'm making a library for another game and I'm for sure not going to use it for that. It causes too many problems for what it's worth. Even the creator of svelte wrote an article on why Typescript is bad. Typescript is amazing, but for libraries it would be good enough to just use JSDocs. What do you think? There would still be type checking with JSDocs but it wouldn't be strict and we'd be able to document how methods work/are intended to work.

I wonder why it would be slowing down so much. My only guess would be it's not actually running in parallel so it's basically the same as before but now it's just taking longer because it has to upload data to the worker? If you use the Chrome devtools and do a CPU, after you stop it you can go to the sources tab and it'll show you which lines took how long to execute

**Comment by Juliiiian** [comment]
Yes, I am 100% on your side with the switch from typescript to JSDocs. I saw the article from Rich Harris, and first I thought it was bad because I just started using Typescript 8 Months ago. But sometimes it's annoying, like when you don't use things like Vite and you need to compile every time you want to run your project. 
I actually used a bit of JSDocs for my Rust Item Steam Price Scraper but at the time I had no idea that it was JSDocs, I just wanted to document my functions.

Yes, that's my guess too, but I had no motivation to look into the code after I got it working or at least the tests passed. I am currently working on a few side projects, and this is more or less my least prioritized project.

**Comment by Dezinater** [comment]
Yeah it's annoying having to build it everytime especially since a library like this can have many small updates. I still need to learn more about JSDocs I've only used it a bit to try it out. Maybe I can make a branch to convert this library soon

What other kind of projects are you working on? I have a few others too which is why I've been slow with this

**Comment by Juliiiian** [comment]
It's mainly motivation for me. I have one side Project it's a gambling site but only for learning to work with web sockets and all that steam stuff (which is absolutely demotivating because the valve docs are no help) but I didn't work on it for half a year on it. For that, I wrote a short script to save all the rust item definitions + requesting the skin prices. And I am currently building a Rust wiki where this will be helpful, so you can preview maps with seeds (at least the plan) for that I copied https://totalrust.com/ into SvelteKit, but now I am looking for a way to extract the newest Item data from the server files, but this will take time. And If I really plan to release the website I need a massive redesign because it looks exactly the same (but worth the SvelteKit experience).

**Comment by Dezinater** [comment]
Haha yea the Valve docs can be lacking. It would be nice to have a wiki where you can preview maps and maybe even see extra info about it. I saw this one rust map site that shows where certain nodes will spawn, which areas have hemp, etc. Maybe that could be good for the wiki? It would basically just be displaying the topologies. For extracting info from the server maybe you should look into game hacking, I'm pretty sure the client and server run the same code so if you reverse the server code all of the client values should be in there as well. I'm pretty sure sites like RustLabs do that to get the latest item info. The items could also just be a Unity asset in the client as a prefab or scriptable object so maybe you can dump the client assets and get it from that too. 
It's always good to recreate sites for the experience. And since you've used SvelteKit it should be simpler to redesign it. I think the hard part will just be coming up with a good design. Also I'm sorry for the slow replies on here, if you want to chat quicker you can just add me on Discord, my username should be the same as on here
